soc/xtensa/intel_adsp/ace/power.c

```diff
@@ -373,4 +379,22 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 	z_xt_ints_on(core_desc[cpu].intenable);
 }
 
+#ifdef CONFIG_ARCH_CPU_IDLE_CUSTOM
+
+void arch_cpu_idle(void)
+{
+	sys_trace_idle();
+
+#if CONFIG_ADSP_CPU_DYNAMIC_CLOCK_GATING
+	/* TODO: Add enum for substates (1 -> clock gating) */
+	if (pm_policy_state_lock_is_active(PM_STATE_ACTIVE, 1)) {
+		DSPCS.bootctl[arch_proc_id()].bctl |= DSPBR_BCTL_WAITIPCG;
+	} else {
+		DSPCS.bootctl[arch_proc_id()].bctl &= ~DSPBR_BCTL_WAITIPCG;
+	}
```

Comment:
Another thing, you are just checking substate 1 here, and for all other states / substates you do the same thing, so why have you declared two substates ?

Reply:
@ceolin AFAIU, the other substate is part of https://github.com/zephyrproject-rtos/zephyr/pull/66092 which has additional patches on top of this series but DNM for now as it requires more work. We'd like to move this dynamic clock gating part (that is in this PR) faster to upstream and it has proven to work reliably and brings immediate power savings.

Reply:
that ok, I just don't think using active power state + locking this state is the right way to do this. In the end you just want some variable to tell you what to do when you get idle. I mean, the way it is implemented now works, but it is more complicated than it should be and it is using the API / subsystem in a way that was not intended to be used.

Where this lock is being set btw ?

Reply:
> Where this lock is being set btw ?

Here you can find lock usage on SOF side https://github.com/thesofproject/sof/pull/8423